File: ccmtools/test/CppGenerator/assembly_nested/impl/assembly/assembly.cc
#include <cassert>
#include <iostream>

#include "assembly.h"

using namespace std;
namespace wx = wamas::platform::utils;
    
Assembly::Assembly()
	: state_ (::Components::INACTIVE)
{
}


Assembly::~Assembly()
{
}


void 
Assembly::build()
  throw(::Components::CreateFailure)
{
  throw ::Components::CreateFailure();
}


void 
Assembly::build(wx::SmartPtr< ::Components::CCMObject> facadeComponent)
  throw (::Components::CreateFailure)
{
  int error = 0;
  ::Components::HomeFinder* homeFinder = ::Components::HomeFinder::Instance ();
  
  try 
  {
    // find home ob components
    wx::SmartPtr<BasicTestHome> basicTestHome(dynamic_cast<BasicTestHome*>
       (homeFinder->find_home_by_name("BasicTestHome").ptr()));
    wx::SmartPtr<UserTestHome> userTestHome(dynamic_cast<UserTestHome*>
       (homeFinder->find_home_by_name("UserTestHome").ptr()));

    // create components
    superTest = wx::SmartPtr<SuperTest>(dynamic_cast<SuperTest*>(facadeComponent.ptr()));
    basicTest = basicTestHome->create();
    userTest = userTestHome->create();
    
    //provide facets
    basicType = basicTest->provide_basicType();
    userType = userTest->provide_userType();

    // connect components
    superTest->connect_innerBasicType(basicType);
    superTest->connect_innerUserType(userType);
    
    
  }
  catch (::Components::HomeNotFound) 
  {
    cout << "DEPLOYMENT ERROR: can't find a home!" << endl;
    error = -1;
  }
  catch (::Components::NotImplemented& e) 
  {
    cout << "DEPLOYMENT ERROR: function not implemented: "
	 << e.what () << endl;
    error = -1;
  }
  catch (::Components::InvalidName& e) 
  {
    cout << "DEPLOYMENT ERROR: invalid name during connection: "
	 << e.what () << endl;
    error = -1;
  }
  catch (std::exception& e) 
  {
    cout << "Standard exception: "
	 << e.what () << endl;
    error = -1;
  }
  catch (...) {
    cout << "DEPLOYMENT ERROR: there is something going wrong!" << endl;
    error = -1;
  }
  
  if (error < 0) 
  {
    throw ::Components::CreateFailure();
  } 
  else 
  {
    state_ = ::Components::INSERVICE;
  }
}


void
Assembly::configuration_complete()
{
  basicTest->configuration_complete();
  userTest->configuration_complete();
}


void
Assembly::tear_down()
	  throw (::Components::RemoveFailure)
{
  int error = 0;
  try 
  {
    // disconnect components
    superTest->disconnect_innerBasicType();
    superTest->disconnect_innerUserType();

    // remove components
    basicTest->remove();
    userTest->remove();
  }
  catch (::Components::HomeNotFound) 
  {
    cerr << "TEARDOWN ERROR: can't find a home!" << endl;
    error = -1;
  }
  catch (::Components::NotImplemented& e) 
  {
    cerr << "TEARDOWN ERROR: function not implemented: " << e.what()
	 << endl;
    error = -1;
  }
  catch (...) 
  {
    cerr << "TEARDOWN ERROR: there is something going wrong!" << endl;
    error = -1;
  }
  
  state_ = ::Components::INACTIVE;
}


::Components::AssemblyState 
Assembly::get_state()
{
  return state_;
}

File: ccmtools/test/JavaLocalComponents/component_assembly_nested/impl/Assembly.java
package world;

import Components.AssemblyState;
import Components.CreateFailure;
import Components.RemoveFailure;
import Components.HomeNotFound;
import Components.InvalidConfiguration;
import Components.CCMObject;
import Components.HomeFinder;

import world.SuperTest;
import world.SuperTestHome;
import world.BasicTypeInterface;
import world.UserTypeInterface;

import world.BasicTest;
import world.BasicTestHome;
import world.BasicTypeInterface;

import world.UserTest;
import world.UserTestHome;
import world.UserTypeInterface;

import world.SuperTestHome;
import world.SuperTest;

public class Assembly implements Components.Assembly
{
    private AssemblyState state;

    // Super Component: SuperTest
    private SuperTest superTest;

    private BasicTypeInterface innerBasicType;

    private UserTypeInterface innerUserType;

    // Inner Component: BasicTest
    private BasicTest basicTest;

    private BasicTypeInterface basicType;

    // Inner Component: UserTest
    private UserTest userTest;

    private UserTypeInterface userType;

    public Assembly()
    {
        state = AssemblyState.INACTIVE;
    }

    public void build() throws CreateFailure
    {
        System.out.println("+Assembly.build()");
        throw new CreateFailure();
    }

    public void build(CCMObject facadeComponent) throws CreateFailure
    {

        System.out.println("+Assembly.build(" + facadeComponent + ")");
        int error = 0;
        HomeFinder homeFinder = HomeFinder.instance();

        try
        {
            // find home ob components
            BasicTestHome basicTestHome = (BasicTestHome) homeFinder.find_home_by_name("BasicTestHome");
            UserTestHome userTestHome = (UserTestHome) homeFinder.find_home_by_name("UserTestHome");

            // create components
            superTest = (SuperTest) facadeComponent;
            basicTest = basicTestHome.create();
            userTest = userTestHome.create();

            // provide facets
            basicType = basicTest.provide_basicType();
            userType = userTest.provide_userType();

            // connect components
            superTest.connect_innerBasicType(basicType);
            superTest.connect_innerUserType(userType);

            state = AssemblyState.INSERVICE;
        }
        catch (HomeNotFound e)
        {
            e.printStackTrace();
            throw new CreateFailure();
        }
        catch (Exception e)
        {
            e.printStackTrace();
            throw new CreateFailure();
        }
    }

    public void configuration_complete() throws InvalidConfiguration
    {
        System.out.println("+Assembly.configuration_complete()");
        basicTest.configuration_complete();
        userTest.configuration_complete();
    }

    public void tear_down() throws RemoveFailure
    {
        System.out.println("+Assembly.tear_down()");
        int error = 0;
        try
        {
            // disconnect components
            superTest.disconnect_innerBasicType();
            superTest.disconnect_innerUserType();

            // remove components
            basicTest.remove();
            userTest.remove();

            state = AssemblyState.INACTIVE;
        }
        catch (Exception e)
        {
            e.printStackTrace();
            throw new RemoveFailure();
        }
    }

    public AssemblyState get_state()
    {
        return state;
    }
}
